### PR TITLE
[RFC] API change: Device - Const-ify device driver instances.

### PR DESCRIPTION
**Pull request by tbursztyka**
This is a fix for #27399 

(Part of #22941, take a look if unaware of the device driver model rework)

Making all device instances constant.

Note: ALL checkpatch warnings are due to original code and not from this PR, so nevermind.

### CODE REVIEW

**Comment by pabigot** [comment]
Existing code that passes a device pointer as an argument to a function that receives a `void *` parameter is going to be a huge problem.  The `const` qualifier must be cast away in the ISR infrastructure, as well as many thread entry functions and other callback user-data parameters.  At this time the necessary conversions don't appear to be automatable.

**Comment by tbursztyka** [comment]
Modules need to be updated now: tee/debug/segger and crypto/mbedtls
That's why CI is failing at the moment.

**Comment by nzmichaelh** [comment]
Hi @tbursztyka sorry if this has already been discussed, but is it OK to split this into multiple smaller PRs?  This is 10k lines of change which is hard to review and hard to merge without conflicts.

For example, could all of the UART related changes be split into a PR?

**Comment by tbursztyka** [comment]
@nzmichaelh see #26426 (there are 3 PRs it's based on, listed on topic message of this PR. I'll add the title of each PR)

File: arch/arc/core/arc_smp.c
Comment (pabigot) [inline arch/arc/core/arc_smp.c:111]: Comment is on the commit titled:

    device: Const-ify all device driver instance pointers

which produced this change.  The provided Coccinelle script:

```
@r_const_dev_basic
  disable optional_qualifier
@
@@
-struct device *
+const struct device
```
is wrong in two ways:
* it changes a pointer to `struct device` to a `const struct device` object
* it fails to recognize a certain style that puts a `const` qualifier on the variable declaration.

For completeness it should be something like this:
```
@
disable optional_qualifier
@
@@
+const
 struct device *

@
disable optional_qualifier
@
@@
+const
 struct device * const
```

File: drivers/kscan/kscan_mchp_xec.c
Comment (pabigot) [inline drivers/kscan/kscan_mchp_xec.c:54]: This looks like a typo.

**Review by pabigot — DISMISSED** [review]
This review excludes the commits related to changing callbacks, which were reviewed separately in their base PRs.

On the whole I think this is good.  But it's also a huge set of changes, and I'm not sure we'll find everything that got broken.  That's ok, we just need to recognize that these problems may occur.

Text changes requested for commits:

* device: Instan~c~*t*iate all device instance*s* as constant

The commit message of

    scripts: coccinelle: Scripts for finding wrong device instance usage

suggests it's still a draft.  I have not yet looked at the details of these as they seem to be used only for the transformation in this PR; not clear if they're intended to be merged.

I think changing the irq_offload and other API to take a `const void *` is a nice solution, though am a bit concerned that casts might be removing the `const` qualifier which risks undefined behavior.  I have not closely examined all the cases to see whether that happens (another thing that could be done with Coccinelle).

File: include/linker/common-ram.ld
Comment (tbursztyka) [reply include/linker/common-ram.ld:28]: I followed the legacy behavior: driver_api is set at build time, and unset if init fails. Could be other way round (though it means it will set the bit for each and every device, where here it unsets if only init fails. probably not a performance hit though)

File: drivers/gpio/gpio_sam0.c
Comment (tbursztyka) [reply drivers/gpio/gpio_sam0.c:170]: yes I had to go through a grep and manually change these afterwards. Not nice at all indeed.

File: arch/arc/core/arc_smp.c
Comment (tbursztyka) [reply arch/arc/core/arc_smp.c:111]: sure

File: drivers/kscan/kscan_mchp_xec.c
Comment (tbursztyka) [reply drivers/kscan/kscan_mchp_xec.c:54]: good point

File: include/linker/common-ram.ld
Comment (tbursztyka) [reply include/linker/common-ram.ld:28]: I'll switch to the other way round, which makes more sense.

**Comment by tbursztyka** [comment]
Ok, PR seems ready.
The are 2 build failures unrelated to that PR, for which I opened issues #27379 and #27463 
Other failures come (time to time) from qemu_arc_hs/qemu_arc_em on tests that never fails on my side. I guess it's due to timing issues on CI so not relevant here.

@carlescufi , @pabigot , @nashif , @galak  (among others) please (re)review.

**Comment by pabigot** [comment]
It looks like the content in patch 5 ("device: Const-ify all device driver instance pointers") has been modified beyond what Coccinelle does.  The rule removes `const` qualifiers were from variable definitions, unlike the approach I proposed, which isn't really an issue though it's an unrelated undocumented change.

But ReStructuredText documentation changes which couldn't have been done by coccinelle are also in it, as well as changes to macros that Coccinelle doesn't see.  Also there are instances of `struct device const` where Coccinelle produces `const struct device`.

Patch 4 description is inaccurate: only two device init functions are changed in it.  It appears to be necessary because the Coccinelle script in patch 5 was not re-invoked on an updated master branch that included those files: if it had that patch would not be necessary.

It would be nicer to have converted callbacks to take `struct device *` instead of `void *` first, where necessary, separate from the switch to `const struct device`.  Then more material could have been handled by Coccinelle.

So the changes I requested were made and my NAK removed, but I'm less comfortable with this patch set than last review because it's not clear what was done automatically and what was done manually. 10K edits to ~1500 files isn't something that can be reviewed when things are mixed this way.

**Comment by carlescufi** [comment]
API meeting 11th August 2020:

The issues that were raised during the meeting can be summarized as:

- Whenever an automated tool is run in order to make changes to the codebase, that set of changes should be the only content in a commit. This will break bisectability, but will also ensure that the changes can be reproduced by a reviewer (as @pabigot indeed did, and spotted this very issue) by running the exact same tool with the same parameters. `git bisect skip` can be used to work around.

**Comment by tbursztyka** [comment]
@pabigot Probably got lost rebase after rebase where I pusch changes in the wrong commit. I'll try to see if I can fix this. (and rebase again against master)

**Comment by pabigot** [comment]
There are still commits in here that cannot be reproduced by invoking the commands identified in the commit message.

I would like to see a plan for coordinating when and how this gets merged relative to other changes in master.  I'd rather not see more failures like #27594.

The only plan I can see being feasible is locking down master and doing a rebase with a pause before each commit that includes automated changes.  At that point run the corresponding automated process and commit the results, then continue the rebase skipping the old commit.  As far as I know that can't be done if automated and manual changes appear in the same commit, because the manual changes will be lost.

While I was reviewing a patch got merged that breaks a rebase in exactly this way.

There are also a few commits that are perhaps worth of being merged, but IMO don't belong as part of this PR as they're not related to the primary goal nor do they appear to interfere with reaching that goal.

Notes on specific commits below.

> [PATCH 05/42] device: Const-ify all device driver instance pointers

still contains changes that are not produced by Coccinelle, though maybe there are fewer. Some are documentation, some appear to be formatting, some are things that Coccinelle couldn't find.

> [PATCH 12/42] drivers: watchdog: Remove useless Kconfig entry

This is clearly a mistake from #11442 where the call should have been updated to `wdog_cmsdk_apb_setup()`.  The patch should be pulled out to a separate PR: it's not related to the purpose of this PR.

> [PATCH 38/42] isr: Normalize usage of device instance through ISR

Contains changes not produced by the script provided in the commit description.

> PATCH 39/42] tests: uart_asyn: Simplify how the UART device is provided

Doesn't seem related to purpose of PR.

> [PATCH 40/42] drivers: Manual const-ification of device driver instance

Whitespace changes not generated by the documented command.

**Comment by andrewboie** [comment]
> > [PATCH 05/42] device: Const-ify all device driver instance pointers
> 
> still contains changes that are not produced by Coccinelle, though maybe there are fewer. Some are documentation, some appear to be formatting, some are things that Coccinelle couldn't find.

I would like to step back a second and ask is this **really** a requirement?

From my point of view, spending tons of time to get coccinelle scripts to catch every last corner case is letting the perfect be the enemy of the good, and presumes that these scripts get re-used far more than the actually do in practice.

**Comment by pabigot** [comment]
> > still contains changes that are not produced by Coccinelle, though maybe there are fewer. Some are documentation, some appear to be formatting, some are things that Coccinelle couldn't find.
> 
> I would like to step back a second and ask is this **really** a requirement?
> 
> From my point of view, spending tons of time to get coccinelle scripts to catch every last corner case is letting the perfect be the enemy of the good, and presumes that these scripts get re-used far more than the actually do in practice.

@andrewboie I agree entirely.  I wouldn't even bother adding the scripts to discover new instances.  I can't see them being used much.

But rebasing this PR is a nightmare as long as the changes they produce are merged with manual changes in the same commit.  You can do it, but you have to hand-edit all the places where automated changes conflict with changes that were merged since the previous base commit.  That hand editing is error prone.  Give it a try.

On the other hand it took 20 minutes to split the 3 commits that have automated changes into two: one that has the exact output of the automation, the other that has the remaining changes.  Half that time was dealing with the conflicts necessary to fix the backslash alignment in long macros, which was manual but straightforward (edit each file, taken the second version of each conflict).  Verify that [the result](https://github.com/pabigot/zephyr/commits/xpr/24873a) matches the original.

Once done, it took four minutes to rebase this on current master by stopping at each commit that's purely automation, discarding it, re-running the automation, and committing the result before continuing.  Most of *that* time was waiting for `spatch` to build the patch file for the first automated change.  Compare [that](https://github.com/pabigot/zephyr/commits/xpr/24873b) against the previous one and verify that the diff is the same as the diff between the master versions the two were based on.

The next rebase should also take about four minutes.  And the one after that.  And so on, for however many rebases are necessary until we can finally get this merged.

At any rate, all I can do is suggest what seems a better way to do this.  It's not my place to insist on it being done that way, and in fact I never did suggest it was a requirement.

**Comment by andrewboie** [comment]
> It's not my place to insist on it being done that way, and in fact I never did suggest it was a requirement.

It's not always perfectly clear in your multi-point reviews which criteria are a gating factor. Glad we cleared that up.

> The next rebase should also take about four minutes. And the one after that. And so on, for however many rebases are necessary until we can finally get this merged.

There are human factors that could be leveraged for this.

Instead of rebasing over and over again, let this air out as-is, and then once there is sufficient review consensus, have the people who are empowered to merge code hold off on merging other conflicting PRs for a day or two until a final rebase and validation pass can happen. In the unlikely event something slips through, fix it in a follow-up. This has worked for us in the past.

**Comment by tbursztyka** [comment]
Opened a new issue on one bug unrelated to this pr: #27634

**Comment by tbursztyka** [comment]
>     [PATCH 05/42] device: Const-ify all device driver instance pointers
>
> still contains changes that are not produced by Coccinelle, though maybe there are fewer. Some are documentation, some appear to be formatting, some are things that Coccinelle couldn't find.

I'll check for the obvious ones, but I won't be exhaustive. (Actually I don't see this as a big issue).

>    [PATCH 12/42] drivers: watchdog: Remove useless Kconfig entry
>
>This is clearly a mistake from #11442 where the call should have been updated to wdog_cmsdk_apb_setup(). The patch should be pulled out to a separate PR: it's not related to the purpose of this PR.

CI has been finding many unrelated issues like this one on this PR. I can pull it out, but CI will again fail on that here. It will again delay over and over this PR...  

>   [PATCH 38/42] isr: Normalize usage of device instance through ISR
>
>Contains changes not produced by the script provided in the commit description.

The commit messages does tell about manual fixes in addition of the cocci stuff.

>    PATCH 39/42] tests: uart_asyn: Simplify how the UART device is provided
>
> Doesn't seem related to purpose of PR.

sure but, let's just keep it there. It's no harm. I'll remove the "Fixes:..." from the commit message though.

>    [PATCH 40/42] drivers: Manual const-ification of device driver instance
>
> Whitespace changes not generated by the documented command.

I guess I did these to avoid the +80chars limit issue.

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c:441]: This variable should be named `dev_data` instead of `driver_data` (that's what is used in the call below and also in the event handler).
Also this `const` here does not seem to bring any value, and it's not consistent with other declarations of such variables, so I'd suggest to remove it.

**Comment by carlescufi** [comment]
@nashif @anangl @henrikbrixandersen @MaureenHelm @mnkp @nordic-krch @galak @dcpleung a review would be appreciated in order to get this into 2.4

**Review by carlescufi — APPROVED** [review]
In general this looks good to me.

The only nit that I wanted to bring up is the placement of the `const` keyword: this PR uses `const struct device *`, but there is also the option of using `struct device const *`, which is equivalent and can be read backwards: "pointer to constant device structure". That said the numbers do favor the current state of the PR:

```
carles@zephyr-nrf ~/src/zephyr/zephyr (master *)
$ git grep "const struct" | wc -l
7896

carles@zephyr-nrf ~/src/zephyr/zephyr (master *)
$ git grep "const \*" | wc -l
174
```

And consistency comes before everything else, so approving this.

**Comment by chrta** [comment]
The following fails to build for me:
```
$ sanitycheck -p native_posix -s tests/kernel/threads/thread_apis/kernel.threads.apis

...

/home/user/projects/zephyrproject/zephyr/tests/kernel/threads/thread_apis/src/test_threads_cancel_abort.c: In function ‘entry_abort_isr’:
/home/user/projects/zephyrproject/zephyr/tests/kernel/threads/thread_apis/src/test_threads_cancel_abort.c:211:14: error: passing argument 1 of ‘irq_offload’ from incompatible pointer type [-Werror=incompatible-pointer-types]
  irq_offload(offload_func, k_current_get());
              ^~~~~~~~~~~~
In file included from /home/user/projects/zephyrproject/zephyr/include/sys/arch_interface.h:34:0,
                 from /home/user/projects/zephyrproject/zephyr/include/arch/cpu.h:12,
                 from /home/user/projects/zephyrproject/zephyr/include/kernel_includes.h:38,
                 from /home/user/projects/zephyrproject/zephyr/include/kernel.h:17,
                 from /home/user/projects/zephyrproject/zephyr/include/zephyr.h:18,
                 from /home/user/projects/zephyrproject/zephyr/subsys/testsuite/ztest/include/ztest.h:58,
                 from /home/user/projects/zephyrproject/zephyr/tests/kernel/threads/thread_apis/src/test_threads_cancel_abort.c:7:
/home/user/projects/zephyrproject/zephyr/include/irq_offload.h:33:6: note: expected ‘irq_offload_routine_t {aka void (*)(const void *)}’ but argument is of type ‘void (*)(void *)’
 void irq_offload(irq_offload_routine_t routine, const void *parameter);
      ^~~~~~~~~~~
cc1: all warnings being treated as errors
```

SDK Version 0.11.1 on ubuntu 18.04

**Comment by tbursztyka** [comment]
@chrta indeed, new additions that I did not see while rebasing

**Review by dcpleung — APPROVED** [review]

File: drivers/ethernet/eth_mcux.c
Comment (MaureenHelm) [inline drivers/ethernet/eth_mcux.c:1018]: this doesn't look right

File: drivers/sensor/lis2ds12/lis2ds12.c
Comment (MaureenHelm) [inline drivers/sensor/lis2ds12/lis2ds12.c:197]: todo?

File: west.yml
Comment (MaureenHelm) [inline west.yml:59]: please don't forget to update these to new SHAs

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
looks good, just a couple nits. thanks for driving this @tbursztyka

File: drivers/sensor/lis2ds12/lis2ds12.c
Comment (tbursztyka) [reply drivers/sensor/lis2ds12/lis2ds12.c:197]: lis2ds12_sample_fetch_temp() does not exist anywhere, and was giving me build error. That could be an externalized patch though

File: drivers/ethernet/eth_mcux.c
Comment (tbursztyka) [reply drivers/ethernet/eth_mcux.c:1018]: indeed

**Review by chrta — APPROVED** [review]

File: drivers/sensor/lis2ds12/lis2ds12.c
Comment (MaureenHelm) [reply drivers/sensor/lis2ds12/lis2ds12.c:197]: @avisconti please review

**Review by henrikbrixandersen — APPROVED** [review]
I have reviewed a good portion of this patch with main focus on the files, I have touched in the past. Looks good to me.

**Review by vanti — APPROVED** [review]
Verified on SimpleLink.

File: west.yml
Comment (tbursztyka) [reply west.yml:59]: I'll update it once these pull requests get merged yes, so a tiny bit before this PR gets merged then.

**Review by MaureenHelm — APPROVED** [review]

**Review by franciscomunoz — APPROVED** [review]

**Comment by carlescufi** [comment]
west.yml SHAs double-checked, merging the PR

**Pull request merged**
